File: runHYSPLIT.py
####################
## PSCF program
## ~runHYSPLIT.py~
## $Rev$
## Sept. 2009
## Satoshi Takahama (stakahama@ucsd.edu)
####################

import os
import time
import operator
import subprocess

## ============== paths/files ======================
Input_file = 'userinputs/input.txt'
Exec_file = './hymodelt.exe'
Meteo_path = './metdata/'
Output_path = './trajectories/'
Output_base = 'tdump-'
Control = 'CONTROL'

## =============== process ======================
def filekey(filename):
    x = filename.split('.')
    return strptime(x[1],'%b%y')[:2] + (int(x[2][1]),)
Meteo_files = sorted(os.listdir(Meteo_path), key=filekey)
Nmet = len(Meteo_files)
Nloc = 1

## ============== function defs ======================
def transform(x,dk=['Start_loc','Start_time','Run_hours',
                    'Vert_coord','Model_top']):
    return dict(zip(dk,[' '.join(x[:3])] + x[3:]))
def label(tm,loc):
    return ('%s-%s' % (tm,str(int(round(float(loc.split()[2]))))))
def time2time(x,fmt='%m/%d/%y %H:%M:%S'):
    return map(lambda x: str(x + 100)[-2:],time.strptime(x,fmt)[:4])
def strip(s,r=[('\r',''),('\n','')]):
    return reduce(lambda x,y: x.replace(y[0],y[1]),r,s)

## ============== read data ======================
infile = [transform(strip(x).split('\t'))
          for i,x in enumerate(open(Input_file,'r')) if i > 0]

## ============  run HYSPLIT ======================
for row in infile[:3]:
    try: os.remove(Control)
    except: pass

    Output_file = ('%s%s' %
                   (Output_base,label('_'.join(time2time(row['Start_time'])),
                                      row['Start_loc'])))
    f = open(Control,'w')
    f.write('\n'.join([' '.join(time2time(row['Start_time'])),
                       str(Nloc),
                       row['Start_loc'],
                       row['Run_hours'],
                       row['Vert_coord'],
                       row['Model_top'],
                       Nmet] +
                      list(reduce(operator.add,zip([Meteo_path]*Nmet,
                                                   Meteo_files))) +
                      [Output_path,Output_file]))
    f.close()
    subprocess.call(Exec_file)

File: runHYSPLIT.r
####################
## PSCF program
## ~runHYSPLIT.r~
## $Rev$
## Sept. 2009
## Satoshi Takahama (stakahama@ucsd.edu)
####################

## ============== paths/files ======================
eval(parse(text=readLines("userinputs/runHYSPLIT_parms.txt")))
addslash <- function(x)
  if(substring(x,nchar(x),nchar(x))=="/") x else file.path(x,"")
Meteo_path <- addslash(Meteo_path)

## =============== process ======================
filekey <- function(filename) {
  x <- do.call(rbind,strsplit(filename,"\\."))
  as.POSIXct(strptime(paste(x[,2],(as.integer(sub("w","",x[,3]))-1)*7+1),
                      "%b%y %d"),tz="GMT")
}
sortmeteo <- function(x) {
  key <- filekey(x)
  x[!is.na(key)][order(na.omit(key))]
}
Meteo_files <- sortmeteo(list.files(Meteo_path))
Nmet <- length(Meteo_files)
Nloc <- 1

## ============== function defs ======================
time2time <- function(x) 
  format(as.POSIXct(x,"%m/%d/%y %T",tz="GMT"),"%y %m %d %H")

## ============== read data ======================
infile <- read.delim(Input_file,colClasses="character")

## ============  run HYSPLIT ======================
for( i in 1:nrow(infile) ) {
  file.remove(Control)
  Output_file <- paste(Output_base,
                       with(infile[i,],
                            sprintf("%s-%.0f",
                                    gsub("[ ]","_",time2time(Time)),
                                    as.numeric(Altitude))),sep="")
  with(infile[i,],
       cat(time2time(Time),                         ## time
           Nloc,                                    ## num loc
           paste(Latitude,Longitude,Altitude),
           Run_hours,
           Vert_coord,
           Model_top,                               ## coord
           Nmet,                                    ## num met
           rbind(rep(Meteo_path,Nmet),Meteo_files), ## met files
           Output_path,                             ## outpath
           Output_file,                             ## outfile
           sep="\n",
           filename=Control))
  system(Exec_file)
}
